### PR TITLE
grpc: add caller info to client and wrap innerClient

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: Ref https://github.com/tikv/pd/issues/8593

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/caller/constant.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	// Component represents the caller component.
+	Component string
+)
+
+// nolint:exported
+const (
+	Unknown   ID = "unknown"
+	Tidb      ID = "tidb"
+	Pd        ID = "pd"
+	Tikv      ID = "tikv"
```
Comment (JmPotato) [inline client/caller/constant.go:29]: I prefer using a more formal name.

```suggestion
	TiDB      ID = "tidb"
	PD        ID = "pd"
	TiKV      ID = "tikv"
```

File: client/caller/constant.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	TiFlash   ID = "tiflash"
+	Lightning ID = "lightning"
+	Dumpling  ID = "dumpling"
+	Br        ID = "br"
+	Dm        ID = "dm"
+	Cdc       ID = "cdc"
```
Comment (JmPotato) [inline client/caller/constant.go:35]: Ditto.

File: client/client.go [hunk 3]
```diff
@@ -141,6 +141,11 @@ type RPCClient interface {
 	// SetExternalTimestamp sets external timestamp
 	SetExternalTimestamp(ctx context.Context, timestamp uint64) error
 
+	// WithCallerID returns a new RPCClient with the specified caller ID.
```
Comment (JmPotato) [inline client/client.go:144]: What about doing a simple explanation here to explain the differences between the called ID and caller component?

File: client/client.go [hunk 4]
```diff
@@ -212,22 +217,10 @@ func (k *serviceModeKeeper) close() {
...
-	updateTokenConnectionCh chan struct{}
+	callerID        caller.ID
+	callerComponent caller.Component
 
-	ctx    context.Context
-	cancel context.CancelFunc
-	wg     sync.WaitGroup
-	tlsCfg *tls.Config
-	option *option
+	inner *innerClient
```
Comment (JmPotato) [inline client/client.go:223]: A nested struct without introducing the field `inner` may reduce the code changes. But I'm not sure whether the pointer here still works, maybe we can have a try.

File: client/caller/constant.go (hunk 1 above)
Comment (okJiang) [reply client/caller/constant.go:29]: For example?

File: client/client.go (hunk 3 above)
Comment (okJiang) [reply client/client.go:144]: - Caller ID can be understood as a binary file; it is a process.  
- Caller component refers to the components within the process.

File: client/client.go (hunk 4 above)
Comment (okJiang) [reply client/client.go:223]: > But I'm not sure whether the pointer here still works, maybe we can have a try.

I tried it before and indeed encountered some issues, so I wrapped the inner client.

File: client/client.go (hunk 3 above)
Comment (JmPotato) [reply client/client.go:144]: I mean write them into the comments.

File: client/client.go (hunk 4 above)
Comment (okJiang) [reply client/client.go:223]: And I think wrapping `innerclient` helps to reduce the amount of code in client.go, as the code in client.go is a bit excessive.

File: client/client.go [hunk 5]
```diff
@@ -141,6 +141,13 @@ type RPCClient interface {
 	// SetExternalTimestamp sets external timestamp
 	SetExternalTimestamp(ctx context.Context, timestamp uint64) error
 
+	// WithCallerID returns a new RPCClient with the specified caller ID.
+	// Caller ID can be understood as a binary file; it is a process.
+	WithCallerID(callerID caller.ID) RPCClient
+	// WithCallerComponent returns a new RPCClient with the specified caller component.
+	// Caller component refers to the components within the process.
+	WithCallerComponent(callerComponent caller.Component) RPCClient
```
Comment (nolouch) [inline client/client.go:149]: Can we provide an automatic method to ingest the CallerID and Component? Like:

```
package processutil

import "os"

// GetCallerID returns the name of the currently running process
func GetCallerID() string {
	return os.Args[0]
}


func GetTheComponent() string {
	// Get the program counter for the calling function
	pc, _, _, ok := runtime.Caller(1)
	if !ok {
		return "unknown"
	}

	// Retrieve the full function name, including the package path
	fullFuncName := runtime.FuncForPC(pc).Name()

	// Find the last slash, which separates the package path from the function name
	lastSlash := strings.LastIndex(fullFuncName, "/")
	if lastSlash == -1 {
		// No slash, use the last period instead, which separates the package and function
		lastSlash = strings.LastIndex(fullFuncName, ".")
	}
	// Extract the package name
	return fullFuncName[:lastSlash]
}

```

Comment (okJiang) [reply client/client.go:149]: Very good suggestion👍, I will consider if there are any flaws

Comment (JmPotato) [reply client/client.go:149]: Perhaps the HTTP client could also adopt this method.

Comment (okJiang) [reply client/client.go:149]: I have added these two functions to this pr. In the default situation, `CreateClient` will set the `callerID` and `callerComponent` by these two functions. I removed `callerID` input from `CreateClient` and keep the `callerComponent` to allow caller set component.

And I expose the upperLayer to allow user to set it. For example, client-go can set it to catch the real caller in TiDB.

How do you think about these changes?

File: client/client.go [hunk 6]
```diff
@@ -440,192 +453,72 @@ func newClientWithKeyspaceName(
...
 	clientCtx, clientCancel := context.WithCancel(ctx)
 	c := &client{
-		keyspaceID:              nullKeyspaceID,
-		updateTokenConnectionCh: make(chan struct{}, 1),
-		ctx:                     clientCtx,
-		cancel:                  clientCancel,
-		svrUrls:                 svrAddrs,
-		tlsCfg:                  tlsCfg,
-		option:                  newOption(),
+		callerID:        caller.GetCallerID(),
```
Comment (nolouch) [inline client/client.go:456]: why not move to innerClient, and make callerComponent as `ClientOption`?

**Review by nolouch — COMMENTED** [review]
rest lgtm

File: client/client.go (hunk 6 above)
Comment (okJiang) [reply client/client.go:456]: What you said is exactly what I thought at the beginning. You can see it [here](https://github.com/tikv/pd/issues/8593#issuecomment-2354705326)

The `ClientOption` is not actually universal; if we want to apply it across all interfaces, we need to further organize the various options. Therefore, I adopted @JmPotato 's suggestion. In my view, I can accept both implementation methods.

Comment (okJiang) [reply client/client.go:456]: Yes, I intend to require the client to set it.

**Comment by okJiang** [comment]
ptal @JmPotato @nolouch 

The conflicts are too much... Let me fix them after your approve😭

File: client/inner_client.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+func (c *innerClient) getOrCreateGRPCConn() (*grpc.ClientConn, error) {
+	cc, err := c.pdSvcDiscovery.GetOrCreateGRPCConn(c.pdSvcDiscovery.getLeaderURL())
+	if err != nil {
+		return nil, err
+	}
+	return cc, err
+}
+
+func (c *innerClient) dispatchTSORequestWithRetry(ctx context.Context) TSFuture {
```
Comment (nolouch) [inline client/inner_client.go:211]: Why have these changes?

File: client/client.go [hunk 8]
```diff
@@ -1597,10 +1437,34 @@ func (c *client) respForErr(observer prometheus.Observer, start time.Time, err e
...
+func (c *client) WithCallerComponent(callerComponent caller.Component) RPCClient {
+	newClient := *c
+	newClient.callerComponent = callerComponent
+	return &newClient
+}
+
+// adjustCallerComponent returns the caller component if it is empty, it
+// is the upper layer of the pd client.
+func adjustCallerComponent(callerComponent caller.Component) caller.Component {
+	if len(callerComponent) != 0 {
```
Comment (JmPotato) [inline client/client.go:1458]: Better trim it first before the check.

Comment (okJiang) [reply client/client.go:1459]: Trim space? https://github.com/tikv/pd/pull/8704/commits/de83c86db37420405deea689e68894c9d39fc2be

**Review by JmPotato — APPROVED** [review]

**Pull request merged**